// langium/lib/grammar/generated/module.js
"use strict";
/******************************************************************************
 * This file was generated by langium-cli 1.0.0.
 * DO NOT EDIT MANUALLY!
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangiumGrammarGeneratedModule = exports.LangiumGrammarGeneratedSharedModule = exports.LangiumGrammarParserConfig = exports.LangiumGrammarLanguageMetaData = void 0;
const ast_1 = require("./ast");
const grammar_1 = require("./grammar");
exports.LangiumGrammarLanguageMetaData = {
    languageId: 'langium',
    fileExtensions: ['.langium'],
    caseInsensitive: false
};
exports.LangiumGrammarParserConfig = {
    maxLookahead: 3,
};
exports.LangiumGrammarGeneratedSharedModule = {
    AstReflection: () => new ast_1.LangiumGrammarAstReflection()
};
exports.LangiumGrammarGeneratedModule = {
    Grammar: () => (0, grammar_1.LangiumGrammarGrammar)(),
    LanguageMetaData: () => exports.LangiumGrammarLanguageMetaData,
    parser: {
        ParserConfig: () => exports.LangiumGrammarParserConfig
    }
};
