// langium/lib/workspace/configuration.js
"use strict";
/******************************************************************************
 * Copyright 2022 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultConfigurationProvider = void 0;
const vscode_languageserver_protocol_1 = require("vscode-languageserver-protocol");
class DefaultConfigurationProvider {
    constructor(services) {
        this.settings = {};
        this.workspaceConfig = false;
        this.initialized = false;
        this.serviceRegistry = services.ServiceRegistry;
        this.connection = services.lsp.Connection;
        services.lsp.LanguageServer.onInitialize(params => {
            var _a, _b;
            this.workspaceConfig = (_b = (_a = params.capabilities.workspace) === null || _a === void 0 ? void 0 : _a.configuration) !== null && _b !== void 0 ? _b : false;
        });
        services.lsp.LanguageServer.onInitialized(_params => {
            var _a;
            const languages = this.serviceRegistry.all;
            (_a = services.lsp.Connection) === null || _a === void 0 ? void 0 : _a.client.register(vscode_languageserver_protocol_1.DidChangeConfigurationNotification.type, {
                // Listen to configuration changes for all languages
                section: languages.map(lang => this.toSectionName(lang.LanguageMetaData.languageId))
            });
        });
    }
    async initialize() {
        if (this.workspaceConfig && this.connection) {
            const languages = this.serviceRegistry.all;
            const configToUpdate = languages.map(lang => { return { section: this.toSectionName(lang.LanguageMetaData.languageId) }; });
            // get workspace configurations (default scope URI)
            const configs = await this.connection.workspace.getConfiguration(configToUpdate);
            configToUpdate.forEach((conf, idx) => {
                this.updateSectionConfiguration(conf.section, configs[idx]);
            });
        }
        this.initialized = true;
    }
    updateConfiguration(change) {
        if (!change.settings) {
            return;
        }
        Object.keys(change.settings).forEach(section => {
            this.updateSectionConfiguration(section, change.settings[section]);
        });
    }
    updateSectionConfiguration(section, configuration) {
        this.settings[section] = configuration;
    }
    async getConfiguration(language, configuration) {
        if (!this.initialized) {
            await this.initialize();
        }
        const sectionName = this.toSectionName(language);
        if (this.settings[sectionName]) {
            return this.settings[sectionName][configuration];
        }
    }
    toSectionName(languageId) {
        return `${languageId}`;
    }
}
exports.DefaultConfigurationProvider = DefaultConfigurationProvider;
