// langium/lib/utils/cst-util.js
"use strict";
/******************************************************************************
 * Copyright 2021 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInteriorNodes = exports.getStartlineNode = exports.getNextNode = exports.getPreviousNode = exports.findLeafNodeAtOffset = exports.isCommentNode = exports.findCommentNode = exports.findDeclarationNodeAtOffset = exports.DefaultNameRegexp = exports.toDocumentSegment = exports.tokenToRange = exports.isCstChildNode = exports.flattenCst = exports.streamCst = void 0;
const syntax_tree_1 = require("../syntax-tree");
const stream_1 = require("./stream");
/**
 * Create a stream of all CST nodes that are directly and indirectly contained in the given root node,
 * including the root node itself.
 */
function streamCst(node) {
    return new stream_1.TreeStreamImpl(node, element => {
        if ((0, syntax_tree_1.isCompositeCstNode)(element)) {
            return element.children;
        }
        else {
            return [];
        }
    }, { includeRoot: true });
}
exports.streamCst = streamCst;
/**
 * Create a stream of all leaf nodes that are directly and indirectly contained in the given root node.
 */
function flattenCst(node) {
    return streamCst(node).filter(syntax_tree_1.isLeafCstNode);
}
exports.flattenCst = flattenCst;
/**
 * Determines whether the specified cst node is a child of the specified parent node.
 */
function isCstChildNode(child, parent) {
    while (child.parent) {
        child = child.parent;
        if (child === parent) {
            return true;
        }
    }
    return false;
}
exports.isCstChildNode = isCstChildNode;
function tokenToRange(token) {
    // Chevrotain uses 1-based indices everywhere
    // So we subtract 1 from every value to align with the LSP
    return {
        start: {
            character: token.startColumn - 1,
            line: token.startLine - 1
        },
        end: {
            character: token.endColumn,
            line: token.endLine - 1
        }
    };
}
exports.tokenToRange = tokenToRange;
function toDocumentSegment(node) {
    const { offset, end, range } = node;
    return {
        range,
        offset,
        end,
        length: end - offset
    };
}
exports.toDocumentSegment = toDocumentSegment;
// The \p{L} regex matches any unicode letter character, i.e. characters from non-english alphabets
// Together with \w it matches any kind of character which can commonly appear in IDs
exports.DefaultNameRegexp = /^[\w\p{L}]$/u;
/**
 * Performs `findLeafNodeAtOffset` with a minor difference: When encountering a character that matches the `nameRegexp` argument,
 * it will instead return the leaf node at the `offset - 1` position.
 *
 * For LSP services, users expect that the declaration of an element is available if the cursor is directly after the element.
 */
function findDeclarationNodeAtOffset(cstNode, offset, nameRegexp = exports.DefaultNameRegexp) {
    if (cstNode) {
        if (offset > 0) {
            const localOffset = offset - cstNode.offset;
            const textAtOffset = cstNode.text.charAt(localOffset);
            if (!nameRegexp.test(textAtOffset)) {
                offset--;
            }
        }
        return findLeafNodeAtOffset(cstNode, offset);
    }
    return undefined;
}
exports.findDeclarationNodeAtOffset = findDeclarationNodeAtOffset;
function findCommentNode(cstNode, commentNames) {
    if (cstNode) {
        const previous = getPreviousNode(cstNode, true);
        if (previous && isCommentNode(previous, commentNames)) {
            return previous;
        }
        if ((0, syntax_tree_1.isRootCstNode)(cstNode)) {
            // Go from the first non-hidden node through all nodes in reverse order
            // We do this to find the comment node which directly precedes the root node
            const endIndex = cstNode.children.findIndex(e => !e.hidden);
            for (let i = endIndex - 1; i >= 0; i--) {
                const child = cstNode.children[i];
                if (isCommentNode(child, commentNames)) {
                    return child;
                }
            }
        }
    }
    return undefined;
}
exports.findCommentNode = findCommentNode;
function isCommentNode(cstNode, commentNames) {
    return (0, syntax_tree_1.isLeafCstNode)(cstNode) && commentNames.includes(cstNode.tokenType.name);
}
exports.isCommentNode = isCommentNode;
function findLeafNodeAtOffset(node, offset) {
    if ((0, syntax_tree_1.isLeafCstNode)(node)) {
        return node;
    }
    else if ((0, syntax_tree_1.isCompositeCstNode)(node)) {
        let firstChild = 0;
        let lastChild = node.children.length - 1;
        while (firstChild <= lastChild) {
            const middleChild = Math.floor((firstChild + lastChild) / 2);
            const n = node.children[middleChild];
            if (n.offset > offset) {
                lastChild = middleChild - 1;
            }
            else if (n.end <= offset) {
                firstChild = middleChild + 1;
            }
            else {
                return findLeafNodeAtOffset(n, offset);
            }
        }
    }
    return undefined;
}
exports.findLeafNodeAtOffset = findLeafNodeAtOffset;
function getPreviousNode(node, hidden = true) {
    while (node.parent) {
        const parent = node.parent;
        let index = parent.children.indexOf(node);
        if (index === 0) {
            node = parent;
        }
        else {
            index--;
            const previous = parent.children[index];
            if (hidden || !previous.hidden) {
                return previous;
            }
        }
    }
    return undefined;
}
exports.getPreviousNode = getPreviousNode;
function getNextNode(node, hidden = true) {
    while (node.parent) {
        const parent = node.parent;
        let index = parent.children.indexOf(node);
        if (parent.children.length - 1 === index) {
            node = parent;
        }
        else {
            index++;
            const next = parent.children[index];
            if (hidden || !next.hidden) {
                return next;
            }
        }
    }
    return undefined;
}
exports.getNextNode = getNextNode;
function getStartlineNode(node) {
    if (node.range.start.character === 0) {
        return node;
    }
    const line = node.range.start.line;
    let last = node;
    let index;
    while (node.parent) {
        const parent = node.parent;
        const selfIndex = index !== null && index !== void 0 ? index : parent.children.indexOf(node);
        if (selfIndex === 0) {
            node = parent;
            index = undefined;
        }
        else {
            index = selfIndex - 1;
            node = parent.children[index];
        }
        if (node.range.start.line !== line) {
            break;
        }
        last = node;
    }
    return last;
}
exports.getStartlineNode = getStartlineNode;
function getInteriorNodes(start, end) {
    const commonParent = getCommonParent(start, end);
    if (!commonParent) {
        return [];
    }
    return commonParent.parent.children.slice(commonParent.a + 1, commonParent.b);
}
exports.getInteriorNodes = getInteriorNodes;
function getCommonParent(a, b) {
    const aParents = getParentChain(a);
    const bParents = getParentChain(b);
    let current;
    for (let i = 0; i < aParents.length && i < bParents.length; i++) {
        const aParent = aParents[i];
        const bParent = bParents[i];
        if (aParent.parent === bParent.parent) {
            current = {
                parent: aParent.parent,
                a: aParent.index,
                b: bParent.index
            };
        }
        else {
            break;
        }
    }
    return current;
}
function getParentChain(node) {
    const chain = [];
    while (node.parent) {
        const parent = node.parent;
        const index = parent.children.indexOf(node);
        chain.push({
            parent,
            index
        });
        node = parent;
    }
    return chain.reverse();
}
