// langium/lib/parser/value-converter.js
"use strict";
/******************************************************************************
 * Copyright 2021 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertBoolean = exports.convertNumber = exports.convertDate = exports.convertBigint = exports.convertInt = exports.convertID = exports.convertString = exports.DefaultValueConverter = void 0;
const ast_1 = require("../grammar/generated/ast");
const internal_grammar_util_1 = require("../grammar/internal-grammar-util");
const grammar_util_1 = require("../utils/grammar-util");
class DefaultValueConverter {
    convert(input, cstNode) {
        let feature = cstNode.feature;
        if ((0, ast_1.isCrossReference)(feature)) {
            feature = (0, grammar_util_1.getCrossReferenceTerminal)(feature);
        }
        if ((0, ast_1.isRuleCall)(feature)) {
            const rule = feature.rule.ref;
            if (!rule) {
                throw new Error('This cst node was not parsed by a rule.');
            }
            return this.runConverter(rule, input, cstNode);
        }
        return input;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    runConverter(rule, input, cstNode) {
        var _a;
        switch (rule.name.toUpperCase()) {
            case 'INT': return convertInt(input);
            case 'STRING': return convertString(input);
            case 'ID': return convertID(input);
            case 'REGEXLITERAL': return convertString(input);
        }
        switch ((_a = (0, internal_grammar_util_1.getRuleType)(rule)) === null || _a === void 0 ? void 0 : _a.toLowerCase()) {
            case 'number': return convertNumber(input);
            case 'boolean': return convertBoolean(input);
            case 'bigint': return convertBigint(input);
            case 'date': return convertDate(input);
            default: return input;
        }
    }
}
exports.DefaultValueConverter = DefaultValueConverter;
function convertString(input) {
    return input.substring(1, input.length - 1);
}
exports.convertString = convertString;
function convertID(input) {
    if (input.charAt(0) === '^') {
        return input.substring(1);
    }
    else {
        return input;
    }
}
exports.convertID = convertID;
function convertInt(input) {
    return parseInt(input);
}
exports.convertInt = convertInt;
function convertBigint(input) {
    return BigInt(input);
}
exports.convertBigint = convertBigint;
function convertDate(input) {
    return new Date(input);
}
exports.convertDate = convertDate;
function convertNumber(input) {
    return Number(input);
}
exports.convertNumber = convertNumber;
function convertBoolean(input) {
    return input.toLowerCase() === 'true';
}
exports.convertBoolean = convertBoolean;
